source/functions/handle-errors.txt

```diff
@@ -0,0 +1,422 @@
+
+- :ref:`Recursively Call Functions in Error Handling Blocks <functions-recursive-call>`
+- :ref:`Use Database Triggers to Retry Functions <functions-database-trigger-retry>`
+
+.. _functions-recursive-call:
+
+Recursively Call Function in Error Handling Blocks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One way you can retry functions is to handle operations that might fail by
+calling a Function :wikipedia:`recursively <Recursion_(computer_science)>`.
+
+On a high-level, this process includes the following components:
+
+- Execute operations that you want to retry in a ``try`` statement
+  and have the Function call itself in a ``catch`` statement.
+- So that the execution doesn't happen indefinitely, set a maximum number of retries.
+  Every time that the Function fails and enters the ``catch`` statement,
+  increment a count of the current number of retries.
+  Stop the recursive execution when the Function's current number of retries
+  reaches the max number of retries.
+- You may also want to `throttle <https://docs.microsoft.com/en-us/azure/architecture/patterns/throttling>`__
+  retries to reduce the total number of executions in a time frame.
+
+The following table describes some advantages and disadvantages
+of handling Function retries with recursive call strategy.
+
+.. list-table::
+   :header-rows: 1
+   :widths: 50 50
+
+   * - Advantages
+     - Disadvantages
+
+   * - - All retry logic occurs within one function.
+       - Function can return a value after retry.
+       - Minimal additional code.
+     - - All retries must occur within a single Function's max execution time.
+
+The following code example demonstrates an implementation of retrying a Function
+by using recursion in error handling blocks.
+
+.. literalinclude:: /examples/generated/functions/mightFail.snippet.recursive-error-handling.js
+   :language: js
+
+.. _functions-database-trigger-retry:
+
+Use Database Triggers to Retry
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can also retry Functions by using a handler Function that manages retry attempts
+with a :ref:`Database Trigger <database-trigger>` to execute retries
+and a MongoDB collection to track previously failed executions.
+
+On a high-level, this process includes the following components:
+
+- **Main Function** that execute the logic you want to retry,
+  wrapped in the handler function (see below bullet point).
+- **Failed execution tracker collection**, a MongoDB collection
+  that tracks failed executions of the main Function.
+- **Handler Function** that invokes the main Function and logs when the function fails
+  to the failed execution tracker collection.
+- **Database Trigger Function** that reruns the handler function whenever
+  the handler function adds an error to the failed execution tracker collection.
+
+You can support multiple main functions with one set of a handler Function,
+execution tracker collection and Database Trigger Function.
```

Comment:
Our style guide [calls for a serial comma](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#commas).
```suggestion
execution tracker collection, and Database Trigger Function.
```